src/libslic3r/PerimeterGenerator.cpp

```diff
@@ -1896,14 +1896,15 @@ void PerimeterGenerator::process_arachne()
         int        loop_number = this->config->wall_loops + surface.extra_perimeters - 1; // 0-indexed loops
         if (this->layer_id == 0 && this->config->only_one_wall_first_layer)
             loop_number = 0;
+
         // BBS: set the topmost layer to be one wall
-        if (loop_number > 0 && config->only_one_wall_top && this->upper_slices == nullptr)
-            loop_number = 0;
+        const bool is_topmost_layer = (loop_number > 0 && this->upper_slices == nullptr) ? true : false;
```

Comment:
loop_number is the wall count, it shouldn't affect is_topmost_layer value.
We shouldn't move back the `loop_number > 0` to the original position.

Reply:
Also, instead of checking the topmost surface only, it's better to check all top surfaces.

Reply:
GitHub doesn't allow me to comment at the exact location.
You can take a look at Line:1926, when we are at that condition, it means top surface, so we should make a copy of input_params and change the is_top_layer to true, then pass the copy to `wallToolPaths`

Reply:
Sounds good, I'll make the change.